Added support for client_credentials grant type

clintonb opened the pull request:
I am adding this functionality so that our services can get access tokens for service accounts. This is preferable to hacking around the issue or creating a long-lived access token. Adapted from caffeinehit/django-oauth2-provider#47.

FYI @nasthagiri @jcdyer @douglashall @rlucioni @jimabramson 

TODO
- [x] Test locally and on sandboxes
- [x] Security review (@nasthagiri you are our resident OAuth 2.0 expert. Any glaring issues here?)

clintonb commented:
FYI @bderusha, this will aid in authentication for Course Discovery's refresh management command.

rlucioni commented:
@clintonb given the nature of this change, I think it would be appropriate to bump the version of this package.

rlucioni commented on provider/views.py:
```diff
@@ -610,6 +618,22 @@ def password(self, request, data, client):
+        scope = data.get('scope')
+
+        if constants.SINGLE_ACCESS_TOKEN:
+                at = self.get_access_token(request, client.user, scope, client)
```
Nitpick: Looks like this is indented too far.

clintonb replied:
Fixed.

rlucioni commented on provider/views.py:
```diff
@@ -610,6 +618,22 @@ def password(self, request, data, client):
+         Handle ``grant_type=client_credentials`` requests as defined in
+         :rfc:`4.4`.
+         """
+        data = self.get_client_credentials_grant(request, data, client)
```
Won't this raise a `NotImplementedError`?

rlucioni replied:
Ignore this please, I'll comment elsewhere.

rlucioni commented on provider/oauth2/views.py:
```diff
@@ -8,13 +8,11 @@
 from provider.utils import now
-from provider.views import AccessToken as AccessTokenView, OAuthError, AccessTokenMixin
-from provider.views import Capture, Authorize, Redirect
+from provider.views import AccessToken as AccessTokenView, OAuthError, AccessTokenMixin, Capture, Authorize, Redirect
```
This (i.e., `import AccessToken as AccessTokenView`) threw me off. It's especially strange considering that a second `AccessTokenView` is being defined in this module.

I understand wanting to be as surgical as possible when it comes to touching this codebase. However, since you've fixed the imports here, would you be opposed to also fixing this so that `AccessToken` isn't renamed on import?

clintonb replied:
Nope. I'm not touching any of that.

jcdyer replied:
An easy fix would be to import views and use views.AccessToken, etc.  Or you could not touch it.

clintonb replied:
I'll take "not touching the code we want to kill" for 400, please, Alex.

I seriously don't even want to make this PR. I would much rather invest the time/energy in getting us to django-oauth-toolkit. This works. It's tested. I'm not touching it. If anyone feels strongly about renaming, submit another PR.

rlucioni commented:
:+1: once version incremented.

clintonb commented:
Having spent a couple hours trying to increase test coverage, specifically focusing on the case when the endpoint should _not_ issue a refresh token, I am giving up. Given that we _do_ use refresh tokens, there is no point in focusing on ensuring that path is covered. The source of the test failure has to do with the `constants.SINGLE_ACCESS_TOKEN` and the fact that I cannot seem to override its value in tests.

nasthagiri commented on provider/oauth2/tests.py:
```diff
@@ -597,6 +597,25 @@ def test_access_token_response_valid_token_type(self):
         token = self._login_authorize_get_token()
         self.assertEqual(token['token_type'], constants.TOKEN_TYPE, token)
 
+    def test_client_credentials_grant(self):
```
Nit: 
1. Separating the tests here into individual test methods.
2. For comprehensiveness, add a test for invalid client_id with client_credentials grant_type.
3. So, in the end, you would have: `test_client_credentials_grant_success`, `test_client_credentials_grant_invalid_secret`, `test_client_credentials_grant_invalid_client_id`.

nasthagiri commented on provider/oauth2/tests.py:
```diff
@@ -597,6 +597,25 @@ def test_access_token_response_valid_token_type(self):
+            'client_secret': self.get_client().client_secret,
+        })
+
+        self.assertEqual(200, response.status_code, response.content)
```
In the success case, verify that:
1. An `access_token` is returned (with an `expires_in` field).
2. A `refresh_token` is NOT returned.

clintonb replied:
See my latest comment regarding testing for refresh tokens. This code base is...not ideal.

nasthagiri replied:
I'm not sure I follow.  Perhaps I missed an earlier conversation on this.  
But the RFC specifically states in Section 4.4.3 for the Client Credentials Grant:

```
A refresh token SHOULD NOT be included.
```

So, if the Client Credentials grant is implemented correctly, then we should make sure in the test that we don't return a `refresh_token`.

nasthagiri replied:
Isn't it as simple as adding a line: `self.assertNotIn('refresh_token', response.data)`?

clintonb replied:
Ah. This is why I wanted someone intimately familiar with the spec reviewing. This code is being adapted from caffeinehit/django-oauth2-provider#47. I admin I haven't fully read the spec, so I was unaware of the need to omit the refresh token. That is great, because testing that has proved frustrating!

nasthagiri replied:
I see.  It's an unmerged PR and still has issues.

nasthagiri commented on provider/oauth2/forms.py:
```diff
@@ -336,3 +336,10 @@ def clean(self):
+    """
+    Validate a client credentials grant request.
+    """
+    scope = ScopeChoiceField(choices=SCOPE_NAMES, required=False)
```
If you are supporting scope, then I believe you need to verify the scope checks out in this form's `clean` method.  For example: https://github.com/edx/django-oauth2-provider/blob/edx/provider/oauth2/forms.py#L260-L272

nasthagiri replied:
If this is an urgent PR and you don't need support for `scope` right now, you can remove the scope field in the form and not worry about this for now.  However, if you choose to do so, I'd recommend removing the scope field from the form since it implies that we support it.

Per RFC:

```
The authorization server MAY fully or partially ignore the scope requested by the client, based on the authorization server policy or the resource owner's instructions.  
```

clintonb replied:
Merging this functionality isn't urgent; but, I also don't want to run the risk of implementing this incorrectly. I am removing this field.

clintonb committed:
De-coupled refresh token creation from access token creation

- Removed creation of refresh token alongside access token
- Mocking constants.SINGLE_ACCESS_TOKEN instead of changing its value, thereby breaking other tests

clintonb committed:
Added support for client credentials grant

clintonb commented:
@nasthagiri it turns out refresh tokens were being created whenever access tokens were created, regardless of whether the refresh token was needed/requested. I have removed that functionality, and updated the tests appropriately. Please take another look.

@rlucioni please take another look as well.

nasthagiri commented on provider/oauth2/views.py:
```diff
@@ -22,12 +20,10 @@ class OAuth2AccessTokenMixin(AccessTokenMixin):
         except AccessToken.DoesNotExist:
             # None found... make a new one!
             at = self.create_access_token(request, user, scope, client)
-            self.create_refresh_token(request, user, scope, at, client)
```
This change makes sense since not all grant types should create a refresh token.
However, what about the grant types that DO require a refresh token?
Specifically, [authorization_code](https://github.com/edx/django-oauth2-provider/blob/edx/provider/views.py#L570) and the non-Public [password](https://github.com/edx/django-oauth2-provider/blob/edx/provider/views.py#L604) grants also call `get_access_token` - expecting to have created a refresh token.

In fact, looking at the code now, there's an unintentional bug with Public clients using the Password Grant with constants.SINGLE_ACCESS_TOKEN enabled - the code generated refresh tokens for them.

My suggestion to you is to use the recommendation in the [Wrong Abstraction](http://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction) article.  Have each of the 3 callers of the `get_access_token` method handle the `DoesNotExist` exception themselves.  So, `authorization_code` and `password` for private Clients will create refresh tokens, while `client_credentials` and `password` for public Clients won't.  (Alternatively, pass a boolean into `get_access_token` - ignoring the article.)

clintonb replied:
`get_access_token` is only used if `constants.SINGLE_ACCESS_TOKEN`. To my knowledge, we do not have that functionality enabled. All of the handlers have a conditional similar to that seen at https://github.com/edx/django-oauth2-provider/blob/edx/provider/views.py#L569-L573. I don't think we (edX) actually use `get_access_token`.

nasthagiri replied:
I see we are making 2 assumptions here:
1. All operators of open edX will never enable `constants.SINGLE_ACCESS_TOKEN`.
2. We will be removing this library in favor of django-oauth-toolkit so this is a temporary fix, anyway.

Not knowing what future lies ahead of us, in the meantime, why have broken functionality in this common library?  My concern is that (A) current (if any open edX instance has enabled constants.SINGLE_ACCESS_TOKEN) and (B) future uses of this library would not be doing the correct thing.

(By the way, the Mobile team was, in fact, considering to enable constants.SINGLE_ACCESS_TOKEN on edx.org since the access control table continues to grow in the multi-device and multi-session use cases.)

So either: (1) remove support for constants.SINGLE_ACCESS_TOKEN entirely - with notification to edx-code, (2) pass in a boolean to this method on whether or not a refresh_token is expected, or (3) refactor the abstraction as suggested above.  I believe, 2 or 3 should be reasonable to implement.  What do you say?

clintonb replied:
I've implemented a combination of 2 and 3. Please have a look at `get_access_and_refresh_tokens()`.

nasthagiri commented on provider/oauth2/tests.py:
```diff
@@ -166,8 +167,8 @@ def test_token_authorization_redirects_to_correct_uri(self):
 
+    @patch('provider.constants.SINGLE_ACCESS_TOKEN', True)
     def test_token_ignores_expired_tokens(self):
-        constants.SINGLE_ACCESS_TOKEN = True
```
Woah.  Can't believe it "patched" the way it did!

nasthagiri commented on provider/oauth2/tests.py:
```diff
@@ -546,7 +536,7 @@ def test_password_grant_confidential(self):
 
         self.assertEqual(200, response.status_code, response.content)
-        self.assertTrue(json.loads(response.content)['refresh_token'])
+        self.assertIn('refresh_token', json.loads(response.content))
```
Please revert this.  The old way ensures that the refresh token is not falsey (`None`, etc).

clintonb replied:
Reverted.

nasthagiri commented:
@clintonb I have completed my latest pass.  This is looking good!  A few remaining comments.

rlucioni commented:
Looks nice @clintonb, :+1:.

nasthagiri commented on provider/views.py:
```diff
@@ -597,16 +631,31 @@ def password(self, request, data, client):
+            'scope': data.get('scope'),
+            'client': client,
+            'create_refresh_token': False,
+        }
```
Doesn't this also need `'reuse_existing_access_token': constants.SINGLE_ACCESS_TOKEN`?

clintonb replied:
Yes. Added.

nasthagiri commented:
:+1: 
Thanks for taking this on @clintonb.  Looks good.  Pending one question.
